h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/SparseMatrixFactory.java

```diff
@@ -0,0 +1,366 @@
+            sparseData[sparseDataRow] = malloc4f(SparseMatrix.MAX_DIM);
+        }
+        if (dataLastRowSize > 0) {
+            sparseData[sparseData.length - 1] = malloc4f(dataLastRowSize);
+        }
+        // Row indices
+        long[][] rowIndices = new long[rowIndicesLastRowSize == 0 ? rowIndicesRowsNumber : rowIndicesRowsNumber + 1][];
+        iterationLimit = rowIndicesLastRowSize == 0 ? rowIndices.length : rowIndices.length - 1;
+        for (int rowIndicesRow = 0; rowIndicesRow < iterationLimit; rowIndicesRow++) {
+            rowIndices[rowIndicesRow] = malloc8(SparseMatrix.MAX_DIM);
+        }
+        if (rowIndicesLastRowSize > 0) {
+            rowIndices[rowIndices.length - 1] = malloc8(rowIndicesLastRowSize);
+        }
+
+        // Column indices
+        int[][] colIndices = new int[colIndicesLastRowSize == 0 ? colIndicesRowsNumber : colIndicesRowsNumber + 1][];
+        iterationLimit = colIndicesLastRowSize == 0 ? colIndices.length : colIndices.length - 1;
+        for (int colIndicesRow = 0; colIndicesRow < iterationLimit; colIndicesRow++) {
+            colIndices[colIndicesRow] = malloc4(SparseMatrix.MAX_DIM);
+        }
+        if (colIndicesLastRowSize > 0) {
+            colIndices[colIndices.length - 1] = malloc4(colIndicesLastRowSize);
+        }
+
+        // Wrap backing arrays into a SparseMatrix object and return them
+        return new SparseMatrix(sparseData, rowIndices, colIndices);
+    }
+
+    protected static SparseMatrixDimensions calculateCSRMatrixDimensions(Chunk[] chunks, DataInfo di, int weightColIndex){
+
+        int[] nonZeroElementsCounts = new int[1];
+        int[] rowIndicesCounts = new int[1];
+
+        for (int i = 0; i < chunks[0].len(); i++) {
+            // Rows with zero weights are going to be ignored
+            if (weightColIndex != -1 && chunks[weightColIndex].atd(i) == 0) continue;
+            rowIndicesCounts[0]++;
+
+            nonZeroElementsCounts[0] += di._cats;
+
+            for (int j = 0; j < di._nums; ++j) {
+                double val = chunks[di._cats + j].atd(i);
+                if (val != 0) {
+                    nonZeroElementsCounts[0]++;
+                }
+            }
+        }
+
+        return new SparseMatrixDimensions(nonZeroElementsCounts, rowIndicesCounts);
+    }
+
+    public static SparseMatrixDimensions calculateCSRMatrixDimensions(Frame f, int[] chunkIds, Vec w, DataInfo di) {
+        CalculateCSRMatrixDimensionsMrFun fun = new CalculateCSRMatrixDimensionsMrFun(f, di, w, chunkIds);
+        H2O.submitTask(new LocalMR(fun, chunkIds.length)).join();
+
+        return new SparseMatrixDimensions(fun._nonZeroElementsCounts, fun._rowIndicesCounts);
+    }
+
+    private static class CalculateCSRMatrixDimensionsMrFun extends MrFun<CalculateCSRMatrixDimensionsMrFun> {
+        private Frame _f;
+        private DataInfo _di;
+        private Vec _w;
+        private int[] _chunkIds;
+
+        // OUT
```

Comment:
Thank you for doing this.

Reply:
credit where credit is due @michalkurka is the original author